tuf/repository/_repository.py

```diff
@@ -0,0 +1,172 @@
+from copy import deepcopy
+from typing import Dict, Generator, Optional, Tuple
+
+from tuf.api.metadata import Metadata, MetaFile, Signed
+
+logger = logging.getLogger(__name__)
+
+
+class AbortEdit(Exception):
+    """Raise to exit the edit() contextmanager without saving changes"""
+
+
+class Repository(ABC):
+    """Abstract class for metadata modifying implementations
+
+    This class is intended to be a base class used in any metadata editing
+    application, whether it is a real repository server or a developer tool.
+
+    Implementations must implement open() and close(), and can then use the
+    edit() contextmanager to implement actual operations.
+
+    A few operations (sign, snapshot and timestamp) are already implemented
+    in this base class.
+    """
+
+    @abstractmethod
+    def open(self, role: str, init: bool = False) -> Metadata:
+        """Load a roles metadata from storage or cache, return it
+
+        If 'init', then create metadata from scratch"""
+        raise NotImplementedError
+
+    @abstractmethod
+    def close(self, role: str, md: Metadata, sign_only: bool = False) -> None:
+        """Write roles metadata into storage
+
+        If sign_only, then just append signatures of all available keys.
+
+        If not sign_only, update expiry and version and replace signatures
+        with ones from all available keys. Keep snapshot_info and targets_infos
+        updated."""
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def targets_infos(self) -> Dict[str, MetaFile]:
+        """Returns the current targets version information
+
+        Not that there is a difference between this and the published snapshot
+        meta: This dictionary reflects the targets metadata that currently
+        exists in the repository, but the dictionary published by snapshot()
+        will also include metadata that no longer exists in the repository.
+        """
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def snapshot_info(self) -> MetaFile:
+        """Returns the information matching current snapshot metadata"""
+        raise NotImplementedError
+
+    @contextmanager
+    def edit(
+        self, role: str, init: bool = False
+    ) -> Generator[Signed, None, None]:
+        """Context manager for editing a roles metadata
+
+        Context manager takes care of loading the roles metadata (or creating
+        new metadata if 'init'), updating expiry and version. The caller can do
+        other changes to the Signed object and when the context manager exits,
+        a new version of the roles metadata is stored.
+
+        Context manager user can raise AbortEdit from inside the with-block to
+        cancel the edit: in this case none of the changes are stored.
+        """
+        md = self.open(role, init)
+        with suppress(AbortEdit):
+            yield md.signed
+            self.close(role, md)
+
+    def sign(self, role: str) -> None:
```

Comment:
no examples here, this really only comes to play with distributed threshold signing: you get metadata from elsewhere, and load it just so that you can sign (without bumping version etc) .